gui/tiletypes.lua

```diff
@@ -0,0 +1,1667 @@
+                z = selectedPos.z,
+            }
+
+            box = Box {
+                self.first_point or selectedPos,
+                self.last_point or (self.first_point and inBoundsMouse or selectedPos)
+            }
+        end
+
+        if box then
+            local avoid_rect = type(self.avoid_rect) == "function" and self.avoid_rect() or self.avoid_rect
+            box:draw(self.tile_map, avoid_rect, self.ascii_fill)
+        end
+    end
+
+    -- Don't call super.onRenderFrame, since this widget should not be drawn
+end
+
+function BoxSelection:hideCursor()
+    -- Hide cursor
+    df.global.game.main_interface.main_designation_selected = df.main_designation_type.NONE
+end
+
+--================================--
+--||        SelectDialog        ||--
+--================================--
+-- Popup for selecting an item from a list, with a search bar
+
+---@class CategoryChoice
+---@field text string|widgets.LabelToken[]
+---@field category string?
+---@field key string?
+---@field item_list (widgets.ListChoice|CategoryChoice)[]
+
+ARROW = string.char(26)
+
+SelectDialog = defclass(SelectDialog, gui.FramedScreen)
+
+SelectDialog.ATTRS{
+    focus_path = "SelectDialog",
+    prompt = "Type or select a item from this list",
+    base_category = "Any item",
+    frame_style = gui.GREY_LINE_FRAME,
+    frame_inset = 1,
+    frame_title = "Select Item",
+    item_list = DEFAULT_NIL, ---@type (widgets.ListChoice|CategoryChoice)[]
+    on_select = DEFAULT_NIL,
+    on_cancel = DEFAULT_NIL,
+    on_close = DEFAULT_NIL,
+}
+
+function SelectDialog:init()
+    self:addviews{
+        widgets.Label{
+            text = {
+                self.prompt, '\n\n',
+                'Category: ', { text = self:cb_getfield('category_str'), pen = COLOR_CYAN }
+            },
+            text_pen = COLOR_WHITE,
+            frame = { l = 0, t = 0 },
+        },
+        widgets.Label{
+            view_id = 'back',
+            visible = false,
+            text = { { key = 'LEAVESCREEN', text = ': Back' } },
+            frame = { r = 0, b = 0 },
+            auto_width = true,
+        },
+        widgets.FilteredList{
+            view_id = 'list',
+            not_found_label = 'No matching items',
+            frame = { l = 0, r = 0, t = 4, b = 2 },
+            icon_width = 2,
+            on_submit = self:callback('onSubmitItem'),
+            edit_on_char=function(c) return c:match('%l') end,
+        },
+        widgets.Label{
+            text = { {
+                key = 'SELECT', text = ': Select',
+                disabled = function() return not self.subviews.list:canSubmit() end
+            } },
+            frame = { l = 0, b = 0 },
+        }
+    }
+
+    self:initCategory(self.base_category, self.item_list)
+end
+
+function SelectDialog:getWantedFrameSize(rect)
+    return math.max(self.frame_width or 40, #self.prompt), math.min(28, rect.height-8)
+end
+
+function SelectDialog:onDestroy()
```

Comment:
it is more common to trigger on_close in `onDismiss`. (see existing dialogs in `gui/dialogs.lua`.) Is there a reason you wanted this (dismissed) screen off the stack first?